### PR TITLE
keymap specific build settings

### PR DESCRIPTION
**Pull request by wez**
Three commits in this PR; the overall effect is to allow specifying per-keymap build options.
This is done via an overrides.mk makefile that can options on or off.
I've also added a build-all script to verify that everything builds.
build-all is the kind of thing we can wire up via Travis CI; it takes a few minutes to build out all combinations of the keyboards and keymaps.

Refs: https://github.com/jackhumbert/qmk_firmware/issues/90

```
$ ./build-all.sh
<lots of output here>
1 failures:
  make -C keyboard/infinity -f Makefile
```

I'll submit a separate PR to fix the infinity build (it's an OS X specific problem)

### CODE REVIEW

**Comment by jackhumbert** [comment]
Thanks for doing this! I'll have to check out the changes in detail later on to make sure it doesn't break anything in the Planck stuff I have setup.

**Comment by ezuk** [comment]
@jackhumbert just a quick ping on this one :) Think we could merge/close this week maybe?

**Comment by jackhumbert** [comment]
Adding `quantum/keymap_extras/` to the path should take care of a lot of the full path includes in a couple of files - I remember doing this in some branch, but I can't remember if I got it merged to master yet. 

I like the idea of the two state system (vs two + undef), but I'm gonna miss the ease of `#ifdef BLAH` - what are your thoughts, @ezuk?

File: keyboard/ergodox_ez/Makefile [hunk 1]
```diff
@@ -88,19 +88,19 @@ OPT_DEFS += -DBOOTLOADER_SIZE=512
 
 
 # Build Options
-#   comment out to disable the options.
+#   Set to yes to enable by default, or make MIDI_ENABLE=yes or
+#   make SLEEP_LED_ENABLE=no to override for a given make invocation
```
Comment (ezuk) [inline keyboard/ergodox_ez/Makefile]: Not sure i follow this comment; care to rephrase maybe? Do you mean to say commenting out no longer works basically, but I need to specify yes/no? What happens if I do comment it out?

File: build-all.sh [hunk 2]
```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env bash
```
Comment (ezuk) [inline build-all.sh]: This is quite cool. Not a direct comment on this one, but just a thought: Would be nice to have a `util` directory with all of the shell scripts we have strewn around the project. The ErgoDox EZ has a `makeallhex.sh` for example, which does something similar but not quite like this.

**Comment by ezuk** [comment]
@jackhumbert my thoughts are on the noobish side of the spectrum, and I left my sheepish questions as line comments. :tada:  :P

File: keyboard/ergodox_ez/Makefile [hunk 3]
```diff
@@ -88,19 +88,19 @@ OPT_DEFS += -DBOOTLOADER_SIZE=512
...
-# CONSOLE_ENABLE   = yes # Console for debug(+400)
-COMMAND_ENABLE   = yes # Commands for debug and configuration
-CUSTOM_MATRIX    = yes # Custom matrix file for the ErgoDox EZ
-SLEEP_LED_ENABLE = yes  # Breathing sleep LED during USB suspend
-NKRO_ENABLE      = yes		# USB Nkey Rollover - if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
-# MIDI_ENABLE    = yes 		# MIDI controls
-# UNICODE_ENABLE = yes 		# Unicode
-
+BOOTMAGIC_ENABLE := yes # Virtual DIP switch configuration(+1000)
+MOUSEKEY_ENABLE  := yes # Mouse keys(+4700)
```
Comment (ezuk) [inline keyboard/ergodox_ez/Makefile]: Also, and this thing is hard to Google, what's the difference between straight-up `=` assignment and `:=` assignment?

Comment (J3RN) [inline keyboard/ergodox_ez/Makefile]: @ezuk I just stumbled across this PR. However, my Google-fu turns up [this useful StackOverflow question](http://stackoverflow.com/questions/448910/makefile-variable-assignment). It seems to be a question of lazy/immediate evaluation, but I'm not sure how it specifically applies here.

Comment (ezuk) [inline keyboard/ergodox_ez/Makefile]: Oh, that's nice! Thank you!

**Comment by jackhumbert** [comment]
@wez thanks for paving the way on this! I was trying to do something similar yesterday, and ended with practically the same code on #256 - it would have been much easier to just merge this :) I think I still have some changes to make to match everything you did.

What do you think about updating this PR to just have the build all script for now?

**Pull request closed without merging**